// @antv/g2/lib/geometry/shape/line/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var util_1 = require("@antv/util");
var base_1 = require("../base");
var get_path_points_1 = require("../util/get-path-points");
var get_style_1 = require("../util/get-style");
var path_1 = require("../util/path");
var split_points_1 = require("../util/split-points");
var util_2 = require("./util");
function getShapeAttrs(cfg, smooth, constraint) {
    var isStack = cfg.isStack, connectNulls = cfg.connectNulls, isInCircle = cfg.isInCircle;
    var points = get_path_points_1.getPathPoints(cfg.points, connectNulls); // 根据 connectNulls 值处理 points
    var path = [];
    util_1.each(points, function (eachLinePoints) {
        path = path.concat(getPath(eachLinePoints, isInCircle, isStack, smooth, constraint));
    });
    return tslib_1.__assign(tslib_1.__assign({}, get_style_1.getStyle(cfg, true, false, 'lineWidth')), { path: path });
}
// 单条 path
function getSinglePath(points, isInCircle, smooth, constraint) {
    var path;
    if (!smooth) {
        path = path_1.getLinePath(points, false);
        if (isInCircle) {
            path.push(['Z']);
        }
    }
    else {
        // 直角坐标系下绘制曲线时限制最大值、最小值
        if (isInCircle && points.length) {
            points.push({ x: points[0].x, y: points[0].y });
        }
        path = path_1.getSplinePath(points, false, constraint);
    }
    return path;
}
function getRangePath(points, isInCircle, isStack, smooth, constraint) {
    var topPoints = [];
    var bottomPoints = [];
    util_1.each(points, function (point) {
        var result = split_points_1.splitPoints(point);
        topPoints.push(result[1]); // 上边
        bottomPoints.push(result[0]); // 底边
    });
    var topPath = getSinglePath(topPoints, isInCircle, smooth, constraint);
    var bottomPath = getSinglePath(bottomPoints, isInCircle, smooth, constraint);
    if (isStack) {
        return topPath;
    }
    return topPath.concat(bottomPath);
}
function getPath(points, isInCircle, isStack, smooth, constraint) {
    var first = points[0];
    return util_1.isArray(first.y)
        ? getRangePath(points, isInCircle, isStack, smooth, constraint)
        : getSinglePath(points, isInCircle, smooth, constraint);
}
var LineShapeFactory = base_1.registerShapeFactory('line', {
    defaultShapeType: 'line',
});
// 这里因为代码公用，所以直接全部注册
// 'line' 默认折线；'dot' 点线 ···；'dash' 断线 - - -
util_1.each(['line', 'dot', 'dash', 'smooth'], function (shapeType) {
    base_1.registerShape('line', shapeType, {
        draw: function (cfg, container) {
            var smooth = shapeType === 'smooth';
            var constraint;
            if (smooth) {
                var _a = this.coordinate, start = _a.start, end = _a.end;
                constraint = [
                    [start.x, end.y],
                    [end.x, start.y],
                ];
            }
            var attrs = getShapeAttrs(cfg, smooth, constraint);
            var shape = container.addShape({
                type: 'path',
                attrs: attrs,
                name: 'line',
            });
            return shape;
        },
        getMarker: function (markerCfg) {
            return util_2.getLineMarker(markerCfg, shapeType);
        },
    });
});
exports.default = LineShapeFactory;
