// @antv/g2/lib/interaction/action/active-region.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var util_1 = require("@antv/util");
var graphics_1 = require("../../util/graphics");
var base_1 = tslib_1.__importDefault(require("./base"));
/**
 * 背景框的 Action
 * @ignore
 */
var ActiveRegion = /** @class */ (function (_super) {
    tslib_1.__extends(ActiveRegion, _super);
    function ActiveRegion() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * 显示
     */
    ActiveRegion.prototype.show = function () {
        var view = this.context.view;
        var ev = this.context.event;
        var tooltipItems = view.getTooltipItems({
            x: ev.x,
            y: ev.y,
        });
        if (util_1.isEqual(tooltipItems, this.items)) {
            // 如果拾取数据同上次相同，则不重复绘制
            return;
        }
        this.items = tooltipItems;
        if (tooltipItems.length) {
            var xField_1 = view.getXScale().field;
            var xValue_1 = tooltipItems[0].data[xField_1];
            // 根据 x 对应的值查找 elements
            var elements_1 = [];
            var geometries = view.geometries;
            util_1.each(geometries, function (geometry) {
                var result = geometry.getElementsBy(function (ele) {
                    var eleData = ele.getData();
                    return eleData[xField_1] === xValue_1;
                });
                elements_1 = elements_1.concat(result);
            });
            // 根据 bbox 计算背景框的面积区域
            if (elements_1.length) {
                var firstBBox = elements_1[0].shape.getCanvasBBox();
                var lastBBox = elements_1[elements_1.length - 1].shape.getCanvasBBox();
                var groupBBox_1 = firstBBox;
                util_1.each(elements_1, function (ele) {
                    var bbox = ele.shape.getCanvasBBox();
                    groupBBox_1.x = Math.min(bbox.minX, groupBBox_1.minX);
                    groupBBox_1.y = Math.min(bbox.minY, groupBBox_1.minY);
                    groupBBox_1.width = Math.max(bbox.maxX, groupBBox_1.maxX) - groupBBox_1.x;
                    groupBBox_1.height = Math.max(bbox.maxY, groupBBox_1.maxY) - groupBBox_1.y;
                });
                var backgroundGroup = view.backgroundGroup, coordinateBBox = view.coordinateBBox;
                var coordinate = view.getCoordinate();
                var path = void 0;
                if (coordinate.isRect) {
                    var xScale = view.getXScale();
                    var appendRatio = xScale.isLinear ? 0 : 0.25; // 如果 x 轴是数值类型，如直方图，不需要家额外的宽度
                    var minX = void 0;
                    var minY = void 0;
                    var width = void 0;
                    var height = void 0;
                    if (coordinate.isTransposed) {
                        minX = coordinateBBox.minX;
                        minY = Math.min(lastBBox.minY, firstBBox.minY) - appendRatio * lastBBox.height;
                        width = coordinateBBox.width;
                        height = groupBBox_1.height + appendRatio * 2 * lastBBox.height;
                    }
                    else {
                        minX = Math.min(firstBBox.minX, lastBBox.minX) - appendRatio * firstBBox.width;
                        minY = Math.min(coordinateBBox.minY, firstBBox.minY);
                        width = groupBBox_1.width + appendRatio * 2 * firstBBox.width;
                        height = coordinateBBox.height;
                    }
                    path = [
                        ['M', minX, minY],
                        ['L', minX + width, minY],
                        ['L', minX + width, minY + height],
                        ['L', minX, minY + height],
                        ['Z'],
                    ];
                }
                else {
                    var firstElement = util_1.head(elements_1);
                    var lastElement = util_1.last(elements_1);
                    var startAngle = graphics_1.getAngle(firstElement.getModel(), coordinate).startAngle;
                    var endAngle = graphics_1.getAngle(lastElement.getModel(), coordinate).endAngle;
                    var center = coordinate.getCenter();
                    // @ts-ignore
                    var radius = coordinate.getRadius();
                    var innterRadius = coordinate.innerRadius * radius;
                    path = graphics_1.getSectorPath(center.x, center.y, radius, startAngle, endAngle, innterRadius);
                }
                if (this.regionPath) {
                    this.regionPath.attr('path', path);
                    this.regionPath.show();
                }
                else {
                    this.regionPath = backgroundGroup.addShape({
                        type: 'path',
                        name: 'active-region',
                        capture: false,
                        attrs: {
                            path: path,
                            fill: '#CCD6EC',
                            opacity: 0.3,
                        },
                    });
                }
            }
        }
    };
    /**
     * 隐藏
     */
    ActiveRegion.prototype.hide = function () {
        if (this.regionPath) {
            this.regionPath.hide();
        }
        // this.regionPath = null;
        this.items = null;
    };
    /**
     * 销毁
     */
    ActiveRegion.prototype.destroy = function () {
        this.hide();
        if (this.regionPath) {
            this.regionPath.remove(true);
        }
        _super.prototype.destroy.call(this);
    };
    return ActiveRegion;
}(base_1.default));
exports.default = ActiveRegion;
