// @antv/g2/lib/dependents.js
"use strict";
/* 依赖的模块，在这里统一引入，方便打包优化 */
Object.defineProperty(exports, "__esModule", { value: true });
var g_base_1 = require("@antv/g-base");
exports.Event = g_base_1.Event;
exports.AbstractGroup = g_base_1.AbstractGroup;
exports.AbstractShape = g_base_1.AbstractShape;
// adjust
var factory_1 = require("@antv/adjust/lib/factory");
exports.registerAdjust = factory_1.registerAdjust;
exports.getAdjust = factory_1.getAdjust;
exports.Adjust = factory_1.Adjust;
// attr
var factory_2 = require("@antv/attr/lib/factory");
exports.getAttribute = factory_2.getAttribute;
exports.Attribute = factory_2.Attribute;
var attr_1 = require("@antv/attr");
exports.Color = attr_1.Color;
// coordinate
var coord_1 = require("@antv/coord");
exports.getCoordinate = coord_1.getCoordinate;
exports.registerCoordinate = coord_1.registerCoordinate;
exports.Coordinate = coord_1.Coordinate;
// scale
var scale_1 = require("@antv/scale");
exports.getScale = scale_1.getScale;
exports.registerScale = scale_1.registerScale;
exports.Scale = scale_1.Scale;
// component
var component_1 = require("@antv/component");
exports.Annotation = component_1.Annotation;
exports.Component = component_1.Component;
exports.Crosshair = component_1.Crosshair;
exports.GroupComponent = component_1.GroupComponent;
exports.HtmlComponent = component_1.HtmlComponent;
exports.Slider = component_1.Slider;
// axis
var LineAxis = component_1.Axis.Line, CircleAxis = component_1.Axis.Circle;
exports.LineAxis = LineAxis;
exports.CircleAxis = CircleAxis;
// grid
var LineGrid = component_1.Grid.Line, CircleGrid = component_1.Grid.Circle;
exports.LineGrid = LineGrid;
exports.CircleGrid = CircleGrid;
// legend
var CategoryLegend = component_1.Legend.Category, ContinuousLegend = component_1.Legend.Continuous;
exports.CategoryLegend = CategoryLegend;
exports.ContinuousLegend = ContinuousLegend;
// Tooltip
var HtmlTooltip = component_1.Tooltip.Html;
exports.HtmlTooltip = HtmlTooltip;
